src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -63,68 +73,209 @@ void BatchNorm<InputDataType, OutputDataType>::Reset()
 template<typename InputDataType, typename OutputDataType>
 template<typename eT>
 void BatchNorm<InputDataType, OutputDataType>::Forward(
-    const arma::Mat<eT>& input, arma::Mat<eT>& output)
+    const arma::Mat<eT>& input,
+    arma::Mat<eT>& output)
 {
-  // Mean and variance over the entire training set will be used to compute
-  // the forward pass when deterministic is set to true.
-  if (deterministic)
-  {
-    // Normalize the input and scale and shift the output.
-    output = input.each_col() - runningMean;
-    output.each_col() %= gamma / arma::sqrt(runningVariance / count + eps);
-    output.each_col() += beta;
-  }
-  else
-  {
-    mean = arma::mean(input, 1);
-    variance = arma::var(input, 1, 1);
+  Log::Assert(input.n_rows % size == 0, "Input features must be divisible \
+      by feature maps.");
+
+  const size_t batchSize = input.n_cols;
 
-    // Normalize the input.
-    output = input.each_col() - mean;
-    inputMean = output;
-    output.each_col() /= arma::sqrt(variance + eps);
+  // Set size of output equal to the size of input.
+  output.set_size(arma::size(input));
 
-    // Use Welford method to compute the sample variance and mean.
-    for (size_t i = 0; i < input.n_cols; ++i)
+  // We will calculate minibatch norm on each channel / feature map.
+  if (!deterministic)
+  {
+    // Input size refers to inputWidth * inputHeight.
+    // If inputSize equals one, input corresponds to output from linear layer.
+    const size_t inputSize = input.n_rows / size;
+
+    // Linear output.
+    if (inputSize == 1)
     {
-      count += 1;
+      mean = arma::mean(input, 1);
+      variance = arma::var(input, 1, 1);
+
+      // Normalize the input.
+      output = input.each_col() - mean;
+      inputMean = output;
+      output.each_col() /= arma::sqrt(variance + eps);
+
+      // Use Welford method to compute the sample variance and mean.
```

Comment:
I guess this is only when average is true so, we should have it inside that if statement